File: apps/api/catora_api/db/models/restaurant.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    Text,
    UniqueConstraint,
    Uuid,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from catora_api.db.base import Base, TimestampMixin, UUIDPrimaryKeyMixin, WorkspaceScopedMixin

JSON_DEFAULT = dict
LIST_DEFAULT = list


class RestaurantBrand(UUIDPrimaryKeyMixin, WorkspaceScopedMixin, TimestampMixin, Base):
    __tablename__ = "restaurant_brands"
    __table_args__ = (
        UniqueConstraint("workspace_id", "canonical_key"),
        CheckConstraint(
            "status IN ('active','inactive','retired')",
            name="valid_restaurant_brand_status",
        ),
    )

    catalog_source_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid,
        ForeignKey("catalog_sources.id", ondelete="SET NULL"),
        index=True,
    )
    canonical_key: Mapped[str] = mapped_column(String(500), nullable=False)
    name: Mapped[str] = mapped_column(String(500), nullable=False)
    legal_name: Mapped[str | None] = mapped_column(String(500))
    website_url: Mapped[str | None] = mapped_column(Text)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="active")
    projection_version: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        default="restaurant-domain/v1",
    )
    projection_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


class RestaurantLocation(UUIDPrimaryKeyMixin, WorkspaceScopedMixin, TimestampMixin, Base):
    __tablename__ = "restaurant_locations"
    __table_args__ = (
        UniqueConstraint("workspace_id", "canonical_key"),
        CheckConstraint(
            "status IN ('active','temporarily_closed','inactive','retired')",
            name="valid_restaurant_location_status",
        ),
    )

    brand_id: Mapped[uuid.UUID] = mapped_column(
        Uuid,
        ForeignKey("restaurant_brands.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    catalog_source_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid,
        ForeignKey("catalog_sources.id", ondelete="SET NULL"),
        index=True,
    )
    canonical_key: Mapped[str] = mapped_column(String(500), nullable=False)
    external_location_id: Mapped[str | None] = mapped_column(String(500), index=True)
    name: Mapped[str] = mapped_column(String(500), nullable=False)
    phone: Mapped[str | None] = mapped_column(String(100))
    website_url: Mapped[str | None] = mapped_column(Text)
    ordering_url: Mapped[str | None] = mapped_column(Text)
    address: Mapped[dict[str, object]] = mapped_column(
        JSONB,
        nullable=False,
        default=JSON_DEFAULT,
    )
    latitude: Mapped[Decimal | None] = mapped_column(Numeric(10, 7))
    longitude: Mapped[Decimal | None] = mapped_column(Numeric(10, 7))
    regular_hours: Mapped[dict[str, object]] = mapped_column(
        JSONB,
        nullable=False,
        default=JSON_DEFAULT,
    )
    special_hours: Mapped[list[object]] = mapped_column(
        JSONB,
        nullable=False,
        default=LIST_DEFAULT,
    )
    service_modes: Mapped[list[object]] = mapped_column(
        JSONB,
        nullable=False,
        default=LIST_DEFAULT,
    )
    facilities: Mapped[list[object]] = mapped_column(
        JSONB,
        nullable=False,
        default=LIST_DEFAULT,
    )
    cuisine_types: Mapped[list[object]] = mapped_column(
        JSONB,
        nullable=False,
        default=LIST_DEFAULT,
    )
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="active")
    projection_version: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        default="restaurant-domain/v1",
    )
    projection_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


class RestaurantIdentityAlias(
    UUIDPrimaryKeyMixin,
    WorkspaceScopedMixin,
    TimestampMixin,
    Base,
):
    __tablename__ = "restaurant_identity_aliases"
    __table_args__ = (
        UniqueConstraint(
            "workspace_id",
            "entity_type",
            "entity_id",
            "normalized_alias",
        ),
        CheckConstraint(
            "entity_type IN ('brand','location','service_area','menu','menu_item')",
            name="valid_restaurant_alias_entity_type",
        ),
        CheckConstraint(
            "status IN ('active','retired')",
            name="valid_restaurant_alias_status",
        ),
        Index(
            "ix_restaurant_identity_alias_lookup",
            "workspace_id",
            "entity_type",
            "normalized_alias",
        ),
    )

    entity_type: Mapped[str] = mapped_column(String(40), nullable=False)
    entity_id: Mapped[uuid.UUID] = mapped_column(Uuid, nullable=False)
    alias: Mapped[str] = mapped_column(String(500), nullable=False)
    normalized_alias: Mapped[str] = mapped_column(String(500), nullable=False)
    source_record_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid,
        ForeignKey("source_records.id", ondelete="SET NULL"),
        index=True,
    )
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="active")


class RestaurantServiceArea(
    UUIDPrimaryKeyMixin,
    WorkspaceScopedMixin,
    TimestampMixin,
    Base,
):
    __tablename__ = "restaurant_service_areas"
    __table_args__ = (
        UniqueConstraint("location_id", "canonical_key"),
        CheckConstraint(
            "area_type IN ('city','district','neighborhood','postal_code','polygon','radius')",
            name="valid_restaurant_service_area_type",
        ),
        CheckConstraint(
            "status IN ('active','inactive','retired')",
            name="valid_restaurant_service_area_status",
        ),
    )

    location_id: Mapped[uuid.UUID] = mapped_column(
        Uuid,
        ForeignKey("restaurant_locations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    canonical_key: Mapped[str] = mapped_column(String(500), nullable=False)
    label: Mapped[str] = mapped_column(String(500), nullable=False)
    area_type: Mapped[str] = mapped_column(String(40), nullable=False)
    geometry: Mapped[dict[str, object]] = mapped_column(
        JSONB,
        nullable=False,
        default=JSON_DEFAULT,
    )
    ordering_url: Mapped[str | None] = mapped_column(Text)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="active")


class RestaurantMenu(UUIDPrimaryKeyMixin, WorkspaceScopedMixin, TimestampMixin, Base):
    __tablename__ = "restaurant_menus"
    __table_args__ = (
        UniqueConstraint("workspace_id", "canonical_key"),
        CheckConstraint(
            "status IN ('active','inactive','retired')",
            name="valid_restaurant_menu_status",
        ),
    )

    brand_id: Mapped[uuid.UUID] = mapped_column(
        Uuid,
        ForeignKey("restaurant_brands.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    location_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid,
        ForeignKey("restaurant_locations.id", ondelete="CASCADE"),
        index=True,
    )
    catalog_source_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid,
        ForeignKey("catalog_sources.id", ondelete="SET NULL"),
        index=True,
    )
    canonical_key: Mapped[str] = mapped_column(String(500), nullable=False)
    name: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    currency: Mapped[str | None] = mapped_column(String(3))
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="active")
    available_from: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    available_until: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    source_updated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    projection_hash: Mapped[str] = mapped_column(String(64), nullable=False)


class RestaurantMenuSection(
    UUIDPrimaryKeyMixin,
    WorkspaceScopedMixin,
    TimestampMixin,
    Base,
):
    __tablename__ = "restaurant_menu_sections"
    __table_args__ = (UniqueConstraint("menu_id", "canonical_key"),)

    menu_id: Mapped[uuid.UUID] = mapped_column(
        Uuid,
        ForeignKey("restaurant_menus.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    canonical_key: Mapped[str] = mapped_column(String(500), nullable=False)
    name: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    position: Mapped[int] = mapped_column(Integer, nullable=False, default=0)


class RestaurantMenuItem(
    UUIDPrimaryKeyMixin,
    WorkspaceScopedMixin,
    TimestampMixin,
    Base,
):
    __tablename__ = "restaurant_menu_items"
    __table_args__ = (
        UniqueConstraint("section_id", "canonical_key"),
        CheckConstraint(
            "availability_state IN "
            "('available','unavailable','unknown','not_applicable','conflicting','stale')",
            name="valid_restaurant_menu_item_availability",
        ),
        CheckConstraint(
            "status IN ('active','inactive','retired')",
            name="valid_restaurant_menu_item_status",
        ),
    )

    section_id: Mapped[uuid.UUID] = mapped_column(
        Uuid,
        ForeignKey("restaurant_menu_sections.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    product_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid,
        ForeignKey("products.id", ondelete="SET NULL"),
        index=True,
    )
    canonical_key: Mapped[str] = mapped_column(String(500), nullable=False)
    name: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    price_amount: Mapped[Decimal | None] = mapped_column(Numeric(18, 6))
    currency: Mapped[str | None] = mapped_column(String(3))
    dietary_facts: Mapped[dict[str, object]] = mapped_column(
        JSONB,
        nullable=False,
        default=JSON_DEFAULT,
    )
    allergen_facts: Mapped[dict[str, object]] = mapped_column(
        JSONB,
        nullable=False,
        default=JSON_DEFAULT,
    )
    availability_state: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="unknown",
    )
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="active")
    source_updated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    projection_hash: Mapped[str] = mapped_column(String(64), nullable=False)


class RestaurantModifierGroup(
    UUIDPrimaryKeyMixin,
    WorkspaceScopedMixin,
    TimestampMixin,
    Base,
):
    __tablename__ = "restaurant_modifier_groups"
    __table_args__ = (
        UniqueConstraint("menu_item_id", "canonical_key"),
        CheckConstraint(
            "min_selections >= 0 AND max_selections >= min_selections",
            name="valid_restaurant_modifier_selection_range",
        ),
    )

    menu_item_id: Mapped[uuid.UUID] = mapped_column(
        Uuid,
        ForeignKey("restaurant_menu_items.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    canonical_key: Mapped[str] = mapped_column(String(500), nullable=False)
    name: Mapped[str] = mapped_column(String(500), nullable=False)
    required: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    min_selections: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    max_selections: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    position: Mapped[int] = mapped_column(Integer, nullable=False, default=0)


class RestaurantModifierOption(
    UUIDPrimaryKeyMixin,
    WorkspaceScopedMixin,
    TimestampMixin,
    Base,
):
    __tablename__ = "restaurant_modifier_options"
    __table_args__ = (
        UniqueConstraint("modifier_group_id", "canonical_key"),
        CheckConstraint(
            "availability_state IN ('available','unavailable','unknown','stale')",
            name="valid_restaurant_modifier_option_availability",
        ),
    )

    modifier_group_id: Mapped[uuid.UUID] = mapped_column(
        Uuid,
        ForeignKey("restaurant_modifier_groups.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    canonical_key: Mapped[str] = mapped_column(String(500), nullable=False)
    name: Mapped[str] = mapped_column(String(500), nullable=False)
    price_delta: Mapped[Decimal | None] = mapped_column(Numeric(18, 6))
    currency: Mapped[str | None] = mapped_column(String(3))
    availability_state: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="unknown",
    )
    position: Mapped[int] = mapped_column(Integer, nullable=False, default=0)


class RestaurantOfferOrPromotion(
    UUIDPrimaryKeyMixin,
    WorkspaceScopedMixin,
    TimestampMixin,
    Base,
):
    __tablename__ = "restaurant_offers_promotions"
    __table_args__ = (
        UniqueConstraint("workspace_id", "canonical_key"),
        CheckConstraint(
            "status IN ('scheduled','active','expired','cancelled','unknown')",
            name="valid_restaurant_offer_status",
        ),
        CheckConstraint(
            "brand_id IS NOT NULL OR location_id IS NOT NULL OR menu_item_id IS NOT NULL",
            name="restaurant_offer_has_scope",
        ),
    )

    brand_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid,
        ForeignKey("restaurant_brands.id", ondelete="CASCADE"),
        index=True,
    )
    location_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid,
        ForeignKey("restaurant_locations.id", ondelete="CASCADE"),
        index=True,
    )
    menu_item_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid,
        ForeignKey("restaurant_menu_items.id", ondelete="CASCADE"),
        index=True,
    )
    canonical_key: Mapped[str] = mapped_column(String(500), nullable=False)
    name: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    starts_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    ends_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="unknown")
    projection_hash: Mapped[str] = mapped_column(String(64), nullable=False)


class RestaurantFreshnessPolicy(
    UUIDPrimaryKeyMixin,
    WorkspaceScopedMixin,
    TimestampMixin,
    Base,
):
    __tablename__ = "restaurant_freshness_policies"
    __table_args__ = (
        UniqueConstraint("workspace_id", "entity_type", "fact_key", "policy_version"),
        CheckConstraint(
            "entity_type IN ('brand','location','service_area','menu','menu_item','offer')",
            name="valid_restaurant_freshness_entity_type",
        ),
        CheckConstraint(
            "warning_age_seconds >= 0 AND max_age_seconds > warning_age_seconds",
            name="valid_restaurant_freshness_ages",
        ),
    )

    entity_type: Mapped[str] = mapped_column(String(40), nullable=False)
    fact_key: Mapped[str] = mapped_column(String(160), nullable=False)
    warning_age_seconds: Mapped[int] = mapped_column(Integer, nullable=False)
    max_age_seconds: Mapped[int] = mapped_column(Integer, nullable=False)
    policy_version: Mapped[str] = mapped_column(String(100), nullable=False)
    rationale: Mapped[str | None] = mapped_column(Text)


class RestaurantFactObservation(
    UUIDPrimaryKeyMixin,
    WorkspaceScopedMixin,
    TimestampMixin,
    Base,
):
    __tablename__ = "restaurant_fact_observations"
    __table_args__ = (
        UniqueConstraint(
            "source_record_id",
            "entity_type",
            "entity_id",
            "fact_key",
            "checksum",
        ),
        CheckConstraint(
            "entity_type IN ('brand','location','service_area','menu','menu_item','offer')",
            name="valid_restaurant_fact_entity_type",
        ),
        CheckConstraint(
            "fact_state IN "
            "('supported','partial','unsupported','stale','conflicting','inaccessible')",
            name="valid_restaurant_fact_state",
        ),
        CheckConstraint(
            "confidence IN ('high','medium','low')",
            name="valid_restaurant_fact_confidence",
        ),
        Index(
            "ix_restaurant_fact_current_lookup",
            "workspace_id",
            "entity_type",
            "entity_id",
            "fact_key",
            "observed_at",
        ),
    )

    source_record_id: Mapped[uuid.UUID] = mapped_column(
        Uuid,
        ForeignKey("source_records.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    entity_type: Mapped[str] = mapped_column(String(40), nullable=False)
    entity_id: Mapped[uuid.UUID] = mapped_column(Uuid, nullable=False)
    fact_key: Mapped[str] = mapped_column(String(160), nullable=False)
    value: Mapped[dict[str, object] | list[object] | str | int | float | bool | None] = (
        mapped_column(JSONB)
    )
    value_type: Mapped[str] = mapped_column(String(50), nullable=False)
    unit: Mapped[str | None] = mapped_column(String(30))
    locale: Mapped[str | None] = mapped_column(String(35))
    fact_state: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="supported",
    )
    confidence: Mapped[str] = mapped_column(String(20), nullable=False, default="high")
    observed_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    effective_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    expires_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    invalidated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    policy_version: Mapped[str | None] = mapped_column(String(100))
    checksum: Mapped[str] = mapped_column(String(64), nullable=False)

File: packages/contracts/src/restaurant.ts
import { z } from "zod";

import { CatalogBridgeJsonValueSchema } from "./catalog-bridge";

export const RESTAURANT_DOMAIN_VERSION = "restaurant-domain/v1" as const;

export const RestaurantFactStateSchema = z.enum([
  "supported",
  "partial",
  "unsupported",
  "stale",
  "conflicting",
  "inaccessible",
]);

export const RestaurantConfidenceSchema = z.enum(["high", "medium", "low"]);

export const RestaurantAvailabilityStateSchema = z.enum([
  "available",
  "unavailable",
  "unknown",
  "not_applicable",
  "conflicting",
  "stale",
]);

export const RestaurantAddressSchema = z
  .object({
    streetAddress: z.string().max(500).nullable().optional(),
    addressLocality: z.string().max(200).nullable().optional(),
    addressRegion: z.string().max(200).nullable().optional(),
    postalCode: z.string().max(50).nullable().optional(),
    addressCountry: z.string().max(100).nullable().optional(),
  })
  .strict();

export const RestaurantGeoCoordinatesSchema = z
  .object({
    latitude: z.number().min(-90).max(90),
    longitude: z.number().min(-180).max(180),
  })
  .strict();

export const RestaurantOpeningHoursIntervalSchema = z
  .object({
    dayOfWeek: z.enum([
      "monday",
      "tuesday",
      "wednesday",
      "thursday",
      "friday",
      "saturday",
      "sunday",
    ]),
    opens: z.string().min(1).max(20),
    closes: z.string().min(1).max(20),
  })
  .strict();

export const RestaurantSpecialHoursIntervalSchema = z
  .object({
    startsAt: z.string().datetime(),
    endsAt: z.string().datetime(),
    opens: z.string().min(1).max(20).nullable().optional(),
    closes: z.string().min(1).max(20).nullable().optional(),
    closed: z.boolean().default(false),
  })
  .strict()
  .superRefine((interval, context) => {
    if (Date.parse(interval.endsAt) <= Date.parse(interval.startsAt)) {
      context.addIssue({
        code: "custom",
        path: ["endsAt"],
        message: "endsAt must be after startsAt",
      });
    }
    if (interval.closed && (interval.opens || interval.closes)) {
      context.addIssue({
        code: "custom",
        path: ["closed"],
        message: "Closed special hours cannot define opening times",
      });
    }
  });

export const RestaurantFactSchema = z
  .object({
    key: z.string().min(1).max(160),
    value: CatalogBridgeJsonValueSchema.nullable().optional(),
    valueType: z.string().min(1).max(50),
    state: RestaurantFactStateSchema,
    confidence: RestaurantConfidenceSchema.default("high"),
    sourceRecordId: z.string().uuid(),
    fieldPath: z.string().min(1).max(500),
    observedAt: z.string().datetime(),
    effectiveAt: z.string().datetime().nullable().optional(),
    expiresAt: z.string().datetime().nullable().optional(),
    invalidatedAt: z.string().datetime().nullable().optional(),
    unit: z.string().max(30).nullable().optional(),
    locale: z.string().max(35).nullable().optional(),
    checksum: z.string().regex(/^[0-9a-f]{64}$/),
    policyVersion: z.string().max(100).nullable().optional(),
  })
  .strict()
  .superRefine((fact, context) => {
    if (fact.state === "supported" && fact.value == null) {
      context.addIssue({
        code: "custom",
        path: ["value"],
        message: "Supported facts require a value",
      });
    }
    if (fact.expiresAt && Date.parse(fact.expiresAt) < Date.parse(fact.observedAt)) {
      context.addIssue({
        code: "custom",
        path: ["expiresAt"],
        message: "expiresAt cannot precede observedAt",
      });
    }
    if (
      fact.invalidatedAt &&
      Date.parse(fact.invalidatedAt) < Date.parse(fact.observedAt)
    ) {
      context.addIssue({
        code: "custom",
        path: ["invalidatedAt"],
        message: "invalidatedAt cannot precede observedAt",
      });
    }
  });

export const RestaurantFreshnessPolicySchema = z
  .object({
    entityType: z.enum([
      "brand",
      "location",
      "service_area",
      "menu",
      "menu_item",
      "offer",
    ]),
    factKey: z.string().min(1).max(160),
    warningAgeSeconds: z.number().int().nonnegative(),
    maxAgeSeconds: z.number().int().positive(),
    policyVersion: z.string().min(1).max(100),
    rationale: z.string().max(2_000).nullable().optional(),
  })
  .strict()
  .superRefine((policy, context) => {
    if (policy.maxAgeSeconds <= policy.warningAgeSeconds) {
      context.addIssue({
        code: "custom",
        path: ["maxAgeSeconds"],
        message: "maxAgeSeconds must exceed warningAgeSeconds",
      });
    }
  });

export const RestaurantIdentityAliasSchema = z
  .object({
    alias: z.string().min(1).max(500),
    normalizedAlias: z.string().min(1).max(500),
    sourceRecordId: z.string().uuid().nullable().optional(),
  })
  .strict();

export const RestaurantServiceAreaSchema = z
  .object({
    canonicalKey: z.string().min(1).max(500),
    label: z.string().min(1).max(500),
    areaType: z.enum([
      "city",
      "district",
      "neighborhood",
      "postal_code",
      "polygon",
      "radius",
    ]),
    geometry: z.record(z.string(), CatalogBridgeJsonValueSchema).default({}),
    orderingUrl: z.string().url().nullable().optional(),
    status: z.enum(["active", "inactive", "retired"]).default("active"),
  })
  .strict();

const MoneyValueSchema = z.string().regex(/^-?\d+(\.\d{1,6})?$/);

export const RestaurantModifierOptionSchema = z
  .object({
    canonicalKey: z.string().min(1).max(500),
    name: z.string().min(1).max(500),
    priceDelta: MoneyValueSchema.nullable().optional(),
    currency: z.string().length(3).nullable().optional(),
    availabilityState: z
      .enum(["available", "unavailable", "unknown", "stale"])
      .default("unknown"),
    position: z.number().int().nonnegative().default(0),
  })
  .strict();

export const RestaurantModifierGroupSchema = z
  .object({
    canonicalKey: z.string().min(1).max(500),
    name: z.string().min(1).max(500),
    required: z.boolean().default(false),
    minSelections: z.number().int().nonnegative().default(0),
    maxSelections: z.number().int().nonnegative().default(1),
    position: z.number().int().nonnegative().default(0),
    options: z.array(RestaurantModifierOptionSchema).max(500).default([]),
  })
  .strict()
  .superRefine((group, context) => {
    if (group.maxSelections < group.minSelections) {
      context.addIssue({
        code: "custom",
        path: ["maxSelections"],
        message: "maxSelections cannot be less than minSelections",
      });
    }
    const keys = new Set<string>();
    for (const [index, option] of group.options.entries()) {
      if (keys.has(option.canonicalKey)) {
        context.addIssue({
          code: "custom",
          path: ["options", index, "canonicalKey"],
          message: "Modifier option canonical keys must be unique",
        });
      }
      keys.add(option.canonicalKey);
    }
  });

export const RestaurantMenuItemSchema = z
  .object({
    canonicalKey: z.string().min(1).max(500),
    name: z.string().min(1).max(500),
    description: z.string().max(100_000).nullable().optional(),
    priceAmount: MoneyValueSchema.nullable().optional(),
    currency: z.string().length(3).nullable().optional(),
    dietaryFacts: z
      .record(z.string().min(1).max(160), CatalogBridgeJsonValueSchema)
      .default({}),
    allergenFacts: z
      .record(z.string().min(1).max(160), CatalogBridgeJsonValueSchema)
      .default({}),
    availabilityState: RestaurantAvailabilityStateSchema.default("unknown"),
    status: z.enum(["active", "inactive", "retired"]).default("active"),
    sourceUpdatedAt: z.string().datetime().nullable().optional(),
    modifiers: z.array(RestaurantModifierGroupSchema).max(100).default([]),
    facts: z.array(RestaurantFactSchema).max(500).default([]),
  })
  .strict();

export const RestaurantMenuSectionSchema = z
  .object({
    canonicalKey: z.string().min(1).max(500),
    name: z.string().min(1).max(500),
    description: z.string().max(20_000).nullable().optional(),
    position: z.number().int().nonnegative().default(0),
    items: z.array(RestaurantMenuItemSchema).max(5_000).default([]),
  })
  .strict();

export const RestaurantMenuSchema = z
  .object({
    canonicalKey: z.string().min(1).max(500),
    name: z.string().min(1).max(500),
    description: z.string().max(20_000).nullable().optional(),
    currency: z.string().length(3).nullable().optional(),
    status: z.enum(["active", "inactive", "retired"]).default("active"),
    availableFrom: z.string().datetime().nullable().optional(),
    availableUntil: z.string().datetime().nullable().optional(),
    sourceUpdatedAt: z.string().datetime().nullable().optional(),
    sections: z.array(RestaurantMenuSectionSchema).max(500).default([]),
    facts: z.array(RestaurantFactSchema).max(500).default([]),
  })
  .strict();

export const RestaurantOfferOrPromotionSchema = z
  .object({
    canonicalKey: z.string().min(1).max(500),
    name: z.string().min(1).max(500),
    description: z.string().max(20_000).nullable().optional(),
    startsAt: z.string().datetime().nullable().optional(),
    endsAt: z.string().datetime().nullable().optional(),
    status: z
      .enum(["scheduled", "active", "expired", "cancelled", "unknown"])
      .default("unknown"),
    facts: z.array(RestaurantFactSchema).max(500).default([]),
  })
  .strict();

export const RestaurantLocationProjectionSchema = z
  .object({
    canonicalKey: z.string().min(1).max(500),
    externalLocationId: z.string().max(500).nullable().optional(),
    name: z.string().min(1).max(500),
    phone: z.string().max(100).nullable().optional(),
    websiteUrl: z.string().url().nullable().optional(),
    orderingUrl: z.string().url().nullable().optional(),
    address: RestaurantAddressSchema.default({}),
    geo: RestaurantGeoCoordinatesSchema.nullable().optional(),
    regularHours: z.array(RestaurantOpeningHoursIntervalSchema).max(100).default([]),
    specialHours: z.array(RestaurantSpecialHoursIntervalSchema).max(500).default([]),
    serviceModes: z
      .array(
        z.enum([
          "dine_in",
          "takeaway",
          "drive_through",
          "delivery",
          "curbside",
          "catering",
        ]),
      )
      .max(20)
      .default([]),
    facilities: z.array(z.string().min(1).max(200)).max(200).default([]),
    cuisineTypes: z.array(z.string().min(1).max(200)).max(100).default([]),
    aliases: z.array(RestaurantIdentityAliasSchema).max(500).default([]),
    serviceAreas: z.array(RestaurantServiceAreaSchema).max(1_000).default([]),
    menus: z.array(RestaurantMenuSchema).max(100).default([]),
    offers: z.array(RestaurantOfferOrPromotionSchema).max(1_000).default([]),
    facts: z.array(RestaurantFactSchema).max(2_000).default([]),
    status: z
      .enum(["active", "temporarily_closed", "inactive", "retired"])
      .default("active"),
  })
  .strict();

export const RestaurantBrandProjectionSchema = z
  .object({
    version: z.literal(RESTAURANT_DOMAIN_VERSION),
    canonicalKey: z.string().min(1).max(500),
    name: z.string().min(1).max(500),
    legalName: z.string().max(500).nullable().optional(),
    websiteUrl: z.string().url().nullable().optional(),
    aliases: z.array(RestaurantIdentityAliasSchema).max(500).default([]),
    locations: z.array(RestaurantLocationProjectionSchema).max(10_000).default([]),
    offers: z.array(RestaurantOfferOrPromotionSchema).max(10_000).default([]),
    facts: z.array(RestaurantFactSchema).max(5_000).default([]),
    status: z.enum(["active", "inactive", "retired"]).default("active"),
  })
  .strict();

export const RestaurantSnapshotSchema = z
  .object({
    version: z.literal(RESTAURANT_DOMAIN_VERSION),
    sourceId: z.string().uuid(),
    snapshotId: z.string().uuid(),
    observedAt: z.string().datetime(),
    brands: z.array(RestaurantBrandProjectionSchema).min(1).max(1_000),
  })
  .strict();

export type RestaurantFactState = z.infer<typeof RestaurantFactStateSchema>;
export type RestaurantFact = z.infer<typeof RestaurantFactSchema>;
export type RestaurantFreshnessPolicy = z.infer<
  typeof RestaurantFreshnessPolicySchema
>;
export type RestaurantMenuItem = z.infer<typeof RestaurantMenuItemSchema>;
export type RestaurantMenu = z.infer<typeof RestaurantMenuSchema>;
export type RestaurantLocationProjection = z.infer<
  typeof RestaurantLocationProjectionSchema
>;
export type RestaurantBrandProjection = z.infer<
  typeof RestaurantBrandProjectionSchema
>;
export type RestaurantSnapshot = z.infer<typeof RestaurantSnapshotSchema>;
